c7n/resources/aws.py

```diff
@@ -157,32 +207,73 @@ def get_bucket_region_clientless(bucket, s3_endpoint):
+        # http error, the status code and response headers provide
+        # context for where the bucket is. For public buckets we
+        # default to raising an exception as unsuitable location at
+        # least for the output use case.
+        #
+        # Dynamic use of urllib trips up static analyzers because of
+        # the potential to accidentally allow unexpected schemes like
+        # file:/. Here we're hardcoding the https scheme, so we can
+        # ignore those specific checks.
+        #
         # nosemgrep: python.lang.security.audit.dynamic-urllib-use-detected.dynamic-urllib-use-detected # noqa
-        response = urlopen(request)  # nosec B310
+        response = url_socket_retry(urlopen, request)  # nosec B310
+        # Successful response indicates a public accessible bucket in the same region
         region = response.headers.get('x-amz-bucket-region')
+
+        if not allow_public:
+            raise ValueError("bucket: '{bucket}' is publicly accessible")
     except HTTPError as err:
-        # Permission errors or redirects for valid buckets should still contain a
-        # header we can use to determine the bucket region.
+        # Returns 404 'Not Found' for buckets that don't exist
+        if err.status == 404:
+            raise ValueError(f"bucket '{bucket}' does not exist")
+        # Permission errors (403) or redirects (301) for valid buckets
+        # should still contain a header we can use to determine the
+        # bucket region. Permission errors are indicative of correct
+        # region, while redirects are for cross region.
         region = err.headers.get('x-amz-bucket-region')
 
     return region
 
 
-def get_bucket_region(bucket, client):
-    """Determine a bucket's region using the GetBucketLocation API action
+def url_socket_retry(func, *args, **kw):
+    """retry a urllib operation in the event of certain errors.
+
+    we want to retry on some common issues for cases where we are
+    connecting through an intermediary proxy or where the downstream
+    is overloaded.
+
+    socket errors
+     - 104 - Connection reset by peer
+     - 110 - Connection timed out
 
-    Look up a bucket's location constraint and map it to a region name as described in
-    https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html
+    http errors
+     - 503 - Slow Down | Service Unavailable
     """
-    location = client.get_bucket_location(Bucket=bucket)['LocationConstraint']
+    min_delay = 1
+    max_delay = 32
+    max_attempts = 4
 
-    # Remap region for cases where the location constraint doesn't match
-    region = {None: 'us-east-1', 'EU': 'eu-west-1'}.get(location, location)
-    return region
+    for idx, delay in enumerate(
+            backoff_delays(min_delay, max_delay, jitter=True)):
+        try:
+            return func(*args, **kw)
+        except HTTPError as err:
+            if not (err.status == 503 and 'Slow Down' in err.reason):
```

Comment:
I've only seen reports of 503s when we hit S3 too hard, but do we need to account for 429 Too Many Requests also?

Reply:
googling around it doesn't seem s3 uses that standard http error code, but instead does a 503 aka service unavailable with the slow down message.

Reply:
fwiw this page has an authoritative list on s3 http err codes 
https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#ErrorCodeList